codetransformer/code.py

```diff
@@ -0,0 +1,432 @@
+        self._cellvars = cellvars
+        self._freevars = freevars
+        self._name = name
+        self._filename = filename
+        self._firstlineno = firstlineno
+        self._lnotab = lnotab
+        self._flags = flags
+
+    @classmethod
+    def from_pycode(cls, co):
+        """Create a Code object from a python code object.
+
+        Parameters
+        ----------
+        co : CodeType
+            The python code object.
+
+        Returns
+        -------
+        code : Code
+            The codetransformer Code object.
+        """
+        # Make it sparse to instrs[n] is the instruction at bytecode[n]
+        instrs = tuple(
+            _sparse_args(
+                Instruction.from_opcode(
+                    b.opcode,
+                    Instruction._no_arg if b.arg is None else b.arg,
+                ) for b in Bytecode(co)
+            ),
+        )
+        for idx, instr in enumerate(instrs):
+            if instr is None:
+                # The sparse value
+                continue
+            if instr.absjmp:
+                instr.arg = instrs[instr.arg]
+            elif instr.reljmp:
+                instr.arg = instrs[instr.arg + idx + 3]
+            elif isinstance(instr, LOAD_CONST):
+                instr.arg = co.co_consts[instr.arg]
+            elif instr.uses_name:
+                instr.arg = co.co_names[instr.arg]
+            elif instr.uses_varname:
+                instr.arg = co.co_varnames[instr.arg]
+
+        flags = co.co_flags
+        has_vargs = bool(flags & Flags.CO_VARARGS)
+        has_kwargs = bool(flags & Flags.CO_VARKEYWORDS)
+
+        # Here we convert the varnames format into our argnames format.
+        paramnames = co.co_varnames[
+            :(co.co_argcount +
+              co.co_kwonlyargcount +
+              has_vargs +
+              has_kwargs)
+        ]
+        # We start with the positional arguments.
+        new_paramnames = list(paramnames[:co.co_argcount])
+        # Add *args next.
+        if has_vargs:
+            new_paramnames.append('*' + paramnames[-1 - has_kwargs])
+        # Add positional only arguments next.
+        new_paramnames.extend(paramnames[
+            co.co_argcount:co.co_argcount + co.co_kwonlyargcount
+        ])
+        # Add **kwargs last.
+        if has_kwargs:
+            new_paramnames.append('**' + paramnames[-1])
+
+        return cls(
+            filter(bool, instrs),
+            argnames=new_paramnames,
+            cellvars=co.co_cellvars,
+            freevars=co.co_freevars,
+            name=co.co_name,
+            filename=co.co_filename,
+            firstlineno=co.co_firstlineno,
+            lnotab=co.co_lnotab,
+            nested=flags & Flags.CO_NESTED,
+            generator=flags & Flags.CO_GENERATOR,
+            coroutine=flags & Flags.CO_COROUTINE,
+            iterable_coroutine=flags & Flags.CO_ITERABLE_COROUTINE
+        )
+
+    @property
+    def as_pycode(self):
```

Comment:
This feels like a method to me.  Partly b/c I don't like properties that do nontrivial computation, but mostly because of the asymmetry with `from_pycode`.